pinot-common/src/main/java/com/linkedin/pinot/common/utils/ServiceStatus.java

```diff
@@ -123,147 +128,130 @@ public String getStatusDescription() {
 
-      if (_helixManager != null) {
-        _helixAdmin = _helixManager.getClusterManagmentTool();
-        _helixDataAccessor = _helixManager.getHelixDataAccessor();
-        _keyBuilder = _helixDataAccessor.keyBuilder();
-      }
-
-      _resourcesToMonitor = resourcesToMonitor;
-
-      LOGGER.info("Monitoring resources {} for start up of instance {}", _resourcesToMonitor, _instanceName);
-    }
+      _resourcesToMonitor = new HashSet<>(resourcesToMonitor);
+      _numTotalResourcesToMonitor = _resourcesToMonitor.size();
 
-    public String getStatusDescription() {
-      return _statusDescription;
+      LOGGER.info("Monitoring {} resources: {} for start up of instance {}", _numTotalResourcesToMonitor,
+          _resourcesToMonitor, _instanceName);
     }
 
     protected abstract T getState(String resourceName);
 
     protected abstract Map<String, String> getPartitionStateMap(T state);
 
     @Override
-    public Status getServiceStatus() {
-      if (_finishedStartingUp) {
+    public synchronized Status getServiceStatus() {
+      if (_resourcesToMonitor.isEmpty()) {
         return Status.GOOD;
       }
-      int index = 0;
-      final int totalResourceCount = _resourcesToMonitor.size();
 
-      for (String resourceToMonitor : _resourcesToMonitor) {
-        final String completedCountStr = "(" + resourceToMonitor + ":" + index + "/" + totalResourceCount + ")";
-        // If the instance is already done starting up, skip checking its state
-        if (_resourcesDoneStartingUp.contains(resourceToMonitor)) {
-          continue;
-        }
-
-        IdealState idealState = getResourceIdealState(resourceToMonitor);
-        T helixState = getState(resourceToMonitor);
-
-        if (idealState == null || helixState == null) {
-          _statusDescription = "idealState or helixState is null" + completedCountStr;
-          return Status.STARTING;
-        }
+      Iterator<String> iterator = _resourcesToMonitor.iterator();
+      while (iterator.hasNext()) {
+        String resourceName = iterator.next();
+        IdealState idealState = getResourceIdealState(resourceName);
 
-        // If the resource is disabled, ignore it
-        if (!idealState.isEnabled()) {
+        // If the resource has been removed or disabled, ignore it
+        if (idealState == null || !idealState.isEnabled()) {
+          iterator.remove();
           continue;
         }
 
-        Map<String, String> statePartitionStateMap = getPartitionStateMap(helixState);
-        if(statePartitionStateMap.isEmpty() && !idealState.getPartitionSet().isEmpty()) {
-          _statusDescription = "statePartitionStateMapSize=" + statePartitionStateMap.size() + ", idealStateSize="
-              + idealState.getPartitionSet().size() + completedCountStr;
+        String descriptionSuffix =
+            String.format("(%s, %d/%d left)", resourceName, _resourcesToMonitor.size(), _numTotalResourcesToMonitor);
```

Comment:
If you log the status as a comma-separate "key=value" format, then we can write scripts to parse the logs.

Reply:
Done